### PR TITLE
ENH: Parse %z and %Z directive in format for to_datetime

### LINKED ISSUE #13486: ENH: bad directive in to_datetime format - this uses std. strptime zone offset
this is really bad for me since almost all linux logfiles contain zone offset.

As far as I can see there is not workaround without parsing the timestamp twice.
#### Code Sample, a copy-pastable example if possible

_parse common Apache access log timestamp_

```
pd.to_datetime(['28/Jul/2006:10:22:04 -0300'], format='%d/%b/%Y:%H:%M:%S %z')

...
/usr/lib/python2.7/dist-packages/pandas/tseries/tools.pyc in _convert_listlike(arg, box, format, name)
    381                 return DatetimeIndex._simple_new(values, name=name, tz=tz)
    382             except (ValueError, TypeError):
--> 383                 raise e
    384 
    385     if arg is None:

ValueError: 'z' is a bad directive in format '%d/%b/%Y:%H:%M:%S %z'
```
#### Expected Output

_should parse timestamps with zone offset_
## INSTALLED VERSIONS

commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-24-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: 1.3.7
pip: 8.1.2
setuptools: 23.0.0
Cython: None
numpy: 1.11.0
scipy: 0.17.0
statsmodels: 0.6.1
IPython: 4.2.0
sphinx: 1.3.6
patsy: 0.4.1
dateutil: 2.4.2
pytz: 2014.10
blosc: None
bottleneck: None
tables: 3.2.2
numexpr: 2.4.3
matplotlib: 1.5.1
openpyxl: 2.3.0
xlrd: 0.9.4
xlwt: 0.7.5
xlsxwriter: None
lxml: 3.5.0
bs4: 4.4.1
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
Jinja2: None

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #13486
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The implimentiontion is near identical to https://github.com/python/cpython/blob/master/Lib/_strptime.py an currently works as `datetime.strptime` would:

```
In [3]: f = '%Y-%m-%d %H:%M:%S %Z %z'

# datetime will parse the case below even though UTC with +0100 should be bogus
In [4]: d = ['2010-01-01 12:00:00 UTC +0100'] * 10

In [5]: pd._libs.tslibs.strptime.array_strptime(np.array(d, dtype='object'), f)
Out[5]:
(array(['2010-01-01T12:00:00.000000000', '2010-01-01T12:00:00.000000000',
        '2010-01-01T12:00:00.000000000', '2010-01-01T12:00:00.000000000',
        '2010-01-01T12:00:00.000000000', '2010-01-01T12:00:00.000000000',
        '2010-01-01T12:00:00.000000000', '2010-01-01T12:00:00.000000000',
        '2010-01-01T12:00:00.000000000', '2010-01-01T12:00:00.000000000'],
       dtype='datetime64[ns]'),
 array([datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC'),
        datetime.timezone(datetime.timedelta(0, 3600), 'UTC')],
       dtype=object))

In [29]: datetime.strptime('2010-01-01 12:00:00 UTC +0100', '%Y-%m-%d %H:%M:%S %Z %z')
Out[29]: datetime.datetime(2010, 1, 1, 12, 0, tzinfo=datetime.timezone(datetime.timedelta(0, 3600), 'UTC'))
```
Currently, an offset needs to get passed (%z) in order for the tzname used be used (%Z).

I'd like to get feedback of what this function should return having parsed %z or %Z. It may be difficult to return a normal DatimeIndex/Series/array given the following edge cases:

1.  User passes strings with different tz offsets (`[date] +0100, [date]. -0600, [date] +1530`)
2.  User passes strings with different tz names (`[date] UTC, [date]. EST, [date] CET`)
3.  User passes strings with incompatable tz name and offset (see example above)

I suppose the most agnostic thing to is to return an array of Timestamps?

```
In [27]: [pd.Timestamp(val, tzinfo=tz) for val, tz in zip(*pd._libs.tslibs.strptime.array_strptime(np.array(d, dtype='object'), f))]
Out[27]:
[Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC'),
 Timestamp('2010-01-01 13:00:00+0100', tz='UTC')]
```

### CODE REVIEW

File: pandas/core/tools/datetimes.py [hunk 1]
```diff
@@ -344,7 +344,7 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
                 if result is None:
                     try:
                         result = array_strptime(arg, format, exact=exact,
-                                                errors=errors)
+                                                errors=errors)[0]
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: why was this changed?

File: pandas/_libs/tslibs/strptime.pyx [hunk 2]
```diff
@@ -281,6 +288,32 @@ def array_strptime(ndarray[object] values, object fmt,
...
+            elif parse_code == 19:
+                z = found_dict['z']
+                if z == 'Z':
+                    gmtoff = 0
+                else:
+                    if z[3] == ':':
+                        z = z[:3] + z[4:]
+                        if len(z) > 5:
+                            if z[5] != ':':
+                                msg = "Unconsistent use of : in {0}"
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: Inconsistent

**Review by jreback — CHANGES_REQUESTED** [review]
need tests

**Comment by mroeschke** [comment]
Summary of the logic thus far and open to feedback:

`array_strptime` will now return 3 arrays of the datetime64 value, tzname, and tzoffset separately.

1) If the tzname directive is only passed, `%Z`, dispatch to `Datimetime`'s tz arg or `Timestamp`'s tz arg if there are multiple different tznames are parsed. 

2) If the tzoffset directive is only passed, `%z`, create a `pytz.FixedOffset` and dispatch to `Datimetime`'s tz arg or `Timestamp`'s tz arg if there are multiple different tzoffset are parsed.

3) If a tzname and tzoffet directives are passed, create a `datetime.timezone` object (similar to the cpython implementation) and just dispatch to `Timestamp`. One question here is if we should invalidate bogus timezones (e.g. UTC +0600, PST -1500, etc)

**Comment by mroeschke** [comment]
I should be able to transfer this logic into cython on the next iteration.

File: pandas/_libs/tslibs/strptime.pyx [hunk 3]
```diff
@@ -281,6 +290,32 @@ def array_strptime(ndarray[object] values, object fmt,
                         else:
                             tz = value
                             break
+            elif parse_code == 19:
+                z = found_dict['z']
+                if z == 'Z':
+                    gmtoff = 0
+                else:
+                    if z[3] == ':':
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: need a check on the len upfront here

File: pandas/_libs/tslibs/strptime.pyx [hunk 4]
```diff
@@ -281,6 +290,32 @@ def array_strptime(ndarray[object] values, object fmt,
...
+                            if z[5] != ':':
+                                msg = "Inconsistent use of : in {0}"
+                                raise ValueError(msg.format(found_dict['z']))
+                            z = z[:5] + z[6:]
+                    hours = int(z[1:3])
+                    minutes = int(z[3:5])
+                    seconds = int(z[5:7] or 0)
+                    gmtoff = (hours * 60 * 60) + (minutes * 60) + seconds
+                    gmtoff_remainder = z[8:]
+                    # Pad to always return microseconds.
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: blank line

File: pandas/_libs/tslibs/strptime.pyx [hunk 5]
```diff
@@ -281,6 +290,32 @@ def array_strptime(ndarray[object] values, object fmt,
                         else:
                             tz = value
                             break
+            elif parse_code == 19:
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: can you move this whole parse to a function and just all it here (and return the values as a tuple)

File: pandas/core/tools/datetimes.py [hunk 6]
```diff
@@ -343,8 +346,74 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
                 # fallback
                 if result is None:
                     try:
-                        result = array_strptime(arg, format, exact=exact,
-                                                errors=errors)
+                        parsing_tzname = '%Z' in format
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: woa, what do you need all this for???

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 7]
```diff
@@ -183,6 +183,63 @@ def test_to_datetime_format_weeks(self, cache):
         for s, format, dt in data:
             assert to_datetime(s, format=format, cache=cache) == dt
 
+    @pytest.mark.skipif(not PY3,
+                        reason="datetime.timezone not supported in PY2")
+    def test_to_datetime_parse_timezone(self):
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: import at the top

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 8]
```diff
@@ -183,6 +183,63 @@ def test_to_datetime_format_weeks(self, cache):
...
+        result = pd.to_datetime(dates, format=fmt)
+        expected_dates = [pd.Timestamp('2010-01-01 12:00:00',
+                                       tzinfo=pytz.FixedOffset(60)),
+                          pd.Timestamp('2010-01-01 12:00:00',
+                                       tzinfo=pytz.FixedOffset(-60))]
+        expected = np.array(expected_dates, dtype=object)
+        tm.assert_numpy_array_equal(result, expected)
+
+        # %z and %Z parsing
+        dates = ['2010-01-01 12:00:00 UTC +0100'] * 2
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: need more checking for invalid (partially formed such as +0, +1foo, UTCbar

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 9]
```diff
@@ -183,6 +183,63 @@ def test_to_datetime_format_weeks(self, cache):
...
+        # %Z parsing only
+        fmt = '%Y-%m-%d %H:%M:%S %Z'
+        dates = ['2010-01-01 12:00:00 UTC'] * 2
+        result = pd.to_datetime(dates, format=fmt)
+        expected_dates = [pd.Timestamp('2010-01-01 12:00:00', tz='UTC')] * 2
+        expected = pd.DatetimeIndex(expected_dates)
+        tm.assert_index_equal(result, expected)
+
+        result = pd.to_datetime(dates, format=fmt, box=False)
+        expected = np.array(expected_dates, dtype=object)
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: use assert_index_equal always

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
@jreback I was able to simplify a lot of my logic ( I underestimated how `Index` would get cast to `DatetimeIndex` when passed `Timestamps` with the same tz).

I created separate functions to parse the timezone directive and then box the result.

File: pandas/_libs/tslibs/strptime.pyx [hunk 10]
```diff
@@ -632,3 +655,48 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
     else:
         days_to_week = week_0_length + (7 * (week_of_year - 1))
         return 1 + days_to_week + day_of_week
+
+cdef _parse_timezone_directive(object z):
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: can be de-privatized (no leading _); these modules are all private.

File: pandas/_libs/tslibs/strptime.pyx [hunk 11]
```diff
@@ -632,3 +655,48 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
...
+    tuple of (offset, offset_fraction)
+    """
+
+    cdef:
+        int gmtoff, gmtoff_fraction, hours, minutes, seconds, pad_number
+        object gmtoff_remainder, gmtoff_remainder_padding
+
+    if z == 'Z':
+        gmtoff = 0
+        gmtoff_fraction = 0
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: would just directly return for this case (0, 0)

then don't need the else

File: pandas/_libs/tslibs/strptime.pyx [hunk 12]
```diff
@@ -632,3 +655,48 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
...
+    else:
+        if z[3] == ':':
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: you might need to wrap this entire block in a try/except if the string is not long enough (or check lengths for each sub-section) and the raise the appropriate error

File: pandas/core/tools/datetimes.py [hunk 13]
```diff
@@ -103,6 +106,70 @@ def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
...
+                               for res, tz in zip(result, tznames)])
+    elif parsing_tzoffset and not parsing_tzname:
+        tz_results = []
+        for res, offset in zip(result, tzoffsets):
+            offset_mins = offset.total_seconds() / 60
+            tzoffset = pytz.FixedOffset(offset_mins)
+            ts = tslib.Timestamp(res)
+            ts = ts.tz_localize(tzoffset)
+            tz_results.append(ts)
+        tz_results = np.array(tz_results)
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: what do you need all of this for, this is jumping thru a lot of hoops here

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 14]
```diff
@@ -25,6 +25,12 @@
 from pandas import (isna, to_datetime, Timestamp, Series, DataFrame,
                     Index, DatetimeIndex, NaT, date_range, compat)
 
+if PY3:
+    from datetime import timezone
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: I think we have a fixture for this

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 15]
```diff
@@ -183,6 +189,71 @@ def test_to_datetime_format_weeks(self, cache):
...
+        ['%Y-%m-%d %H:%M:%S %z',
+         ['2010-01-01 12:00:00 +0100', '2010-01-01 12:00:00 -0100'],
+         [pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(60)),
+          pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(-60))]]])
+    def test_to_datetime_parse_tzname_or_tzoffset(self, box, const,
+                                                  assert_equal, fmt,
+                                                  dates, expected_dates):
+        # %z or %Z parsing
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: add the issue number

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 16]
```diff
@@ -183,6 +189,71 @@ def test_to_datetime_format_weeks(self, cache):
...
+        [['2010-01-01 12:00:00 UTC +0100', '2010-01-01 12:00:00 GMT -0200'],
+         [pd.Timestamp('2010-01-01 13:00:00',
+                       tzinfo=timezone(timedelta(minutes=60), 'UTC')),
+          pd.Timestamp('2010-01-01 10:00:00',
+                       tzinfo=timezone(timedelta(minutes=-120), 'GMT'))]]])
+    def test_to_datetime_parse_tzname_and_tzoffset(self, box, const,
+                                                   assert_equal, dates,
+                                                   expected_dates):
+        # %z and %Z parsing
+        fmt = '%Y-%m-%d %H:%M:%S %Z %z'
```
Comment (jreback) [inline pandas/tests/indexes/datetimes/test_tools.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/indexes/datetimes/test_tools.py (hunk 14 above)
Comment (mroeschke) [reply pandas/tests/indexes/datetimes/test_tools.py]: We have a fixture for `timezone.utc`, but I am testing parsing a custom timezone.

File: pandas/core/tools/datetimes.py (hunk 13 above)
Comment (mroeschke) [reply pandas/core/tools/datetimes.py]: This `elif` branch is:

1) Creating a `pytz.FixedOffset` from the parsed offset
2) Creating a naive Timestamp, then localizing it to the `pytz.FixedOffset` (can't do it directly like `Timezone(res, tz=pytz.FixedOffset(...))` because of my realization from #20257)

File: pandas/_libs/tslibs/strptime.pyx [hunk 17]
```diff
@@ -632,3 +655,48 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
...
+    tuple of (offset, offset_fraction)
+    """
+
+    cdef:
+        int gmtoff, gmtoff_fraction, hours, minutes, seconds, pad_number
+        object gmtoff_remainder, gmtoff_remainder_padding
+
+    if z == 'Z':
+        return (0, 0)
+    else:
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: you don't need the else

can you comment liberally here

File: pandas/_libs/tslibs/strptime.pyx [hunk 18]
```diff
@@ -632,3 +655,48 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
...
+                        raise ValueError(msg.format(z))
+                    z = z[:5] + z[6:]
+            hours = int(z[1:3])
+            minutes = int(z[3:5])
+            seconds = int(z[5:7] or 0)
+            gmtoff = (hours * 60 * 60) + (minutes * 60) + seconds
+            gmtoff_remainder = z[8:]
+        except ValueError:
+            raise ValueError("Could not parse offset: {0}".format(z))
+        # Pad to always return microseconds.
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: blank line

File: pandas/_libs/tslibs/strptime.pyx [hunk 19]
```diff
@@ -632,3 +655,48 @@ cdef _calc_julian_from_U_or_W(int year, int week_of_year,
...
+                        msg = "Inconsistent use of : in {0}"
+                        raise ValueError(msg.format(z))
+                    z = z[:5] + z[6:]
+            hours = int(z[1:3])
+            minutes = int(z[3:5])
+            seconds = int(z[5:7] or 0)
+            gmtoff = (hours * 60 * 60) + (minutes * 60) + seconds
+            gmtoff_remainder = z[8:]
+        except ValueError:
+            raise ValueError("Could not parse offset: {0}".format(z))
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: what exactly is this catching? (other than the exception you are raising)

File: pandas/core/tools/datetimes.py [hunk 20]
```diff
@@ -103,6 +106,70 @@ def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
     return result.values
 
 
+def _return_parsed_timezone_results(result, tznames, tzoffsets, parsing_tzname,
+                                    parsing_tzoffset, box):
+    """
+    Return results from array_strptime if a %z or %Z directive was passed.
+
+    If %Z is only parsed, timezones will be a pytz.timezone object.
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: why are you returning a datetime.timezone at all? we barely use these, prefer pytz.timezone

how is it possible %z and %Z both are parsed?

File: pandas/core/tools/datetimes.py [hunk 21]
```diff
@@ -103,6 +106,70 @@ def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
...
+            offset_mins = offset.total_seconds() / 60
+            tzoffset = pytz.FixedOffset(offset_mins)
+            ts = tslib.Timestamp(res)
+            ts = ts.tz_localize(tzoffset)
+            tz_results.append(ts)
+        tz_results = np.array(tz_results)
+    elif parsing_tzoffset and parsing_tzname:
+        if not PY3:
+            raise ValueError("Parsing tzoffsets are not "
+                             "not supported in Python 3")
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: this should be Python 2

File: pandas/core/tools/datetimes.py [hunk 22]
```diff
@@ -103,6 +106,70 @@ def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
...
+        - Index-like if box=True
+        - ndarray of Timestamps if box=False
+
+    """
+    if parsing_tzname and not parsing_tzoffset:
+        tz_results = np.array([tslib.Timestamp(res, tz=tz)
+                               for res, tz in zip(result, tznames)])
+    elif parsing_tzoffset and not parsing_tzname:
+        tz_results = []
+        for res, offset in zip(result, tzoffsets):
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: you can write this as a list-comprehension

File: pandas/core/tools/datetimes.py [hunk 23]
```diff
@@ -103,6 +106,70 @@ def _convert_and_box_cache(arg, cache_array, box, errors, name=None):
...
+            ts = tslib.Timestamp(res)
+            ts = ts.tz_localize(tzoffset)
+            tz_results.append(ts)
+        tz_results = np.array(tz_results)
+    elif parsing_tzoffset and parsing_tzname:
+        if not PY3:
+            raise ValueError("Parsing tzoffsets are not "
+                             "not supported in Python 3")
+        from datetime import timezone
+        tz_results = []
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: this can be a list-comprehension

File: pandas/core/tools/datetimes.py [hunk 24]
```diff
@@ -343,8 +410,19 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
...
+                        parsing_tzname = '%Z' in format
+                        parsing_tzoffset = '%z' in format
+                        if tz is not None and (parsing_tzname or
+                                               parsing_tzoffset):
+                            raise ValueError("Cannot pass a tz argument when "
+                                             "parsing strings with timezone "
+                                             "information.")
+                        result, tznames, tzoffsets = array_strptime(
+                            arg, format, exact=exact, errors=errors)
+                        if parsing_tzname or parsing_tzoffset:
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: why do you need the parsing_tzoffset & parsing_tzname, won't tznames and tzoffsets be not None in this case? (and None if these are not found)?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
I've changed a couple of things after your review @jreback 

1) Either %z or %Z can be parsed (not together)

2) %z will return a `pytz.FixedOffset`

```
In [3]: pd.to_datetime(['2010-01-01 12:00:00 +0100'], format='%Y-%m-%d %H:%M:%S %z')
Out[3]: DatetimeIndex(['2010-01-01 12:00:00+01:00'], dtype='datetime64[ns, pytz.FixedOffset(60)]', freq=None)
```

3) %Z can parse any timezone specified by `pytz.all_timezones`

```
In [2]: pd.to_datetime(['2010-01-01 12:00:00 US/Pacific'], format='%Y-%m-%d %H:%M:%S %Z')
Out[2]: DatetimeIndex(['2010-01-01 12:00:00-08:00'], dtype='datetime64[ns, US/Pacific]', freq=None)
```

File: pandas/_libs/tslibs/strptime.pyx [hunk 25]
```diff
@@ -58,15 +59,36 @@ def array_strptime(ndarray[object] values, object fmt,
...
+        int week_of_year, week_of_year_start, parse_code
         int64_t us, ns
-        object val, group_key, ampm, found
+        object val, group_key, ampm, found, timezone
         dict found_key
         bint is_raise = errors=='raise'
         bint is_ignore = errors=='ignore'
         bint is_coerce = errors=='coerce'
         int ordinal
+        dict _parse_code_table = {'y': 0,
```
Comment (jreback) [inline pandas/_libs/tslibs/strptime.pyx]: you could make this a module level variable

File: pandas/core/tools/datetimes.py [hunk 26]
```diff
@@ -343,8 +375,20 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
...
+                        parsing_tzname = '%Z' in format
+                        parsing_tzoffset = '%z' in format
+                        if parsing_tzoffset and parsing_tzname:
+                            raise ValueError("Cannot parse both %Z and %z")
+                        elif tz is not None and (parsing_tzname or
+                                                 parsing_tzoffset):
+                            raise ValueError("Cannot pass a tz argument when "
+                                             "parsing strings with timezone "
+                                             "information.")
+                        result, timezones = array_strptime(
```
Comment (jreback) [inline pandas/core/tools/datetimes.py]: I would much rather do the error handling in the _return_parsed_timezone_results. This block is just very complicated and hard to grok

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm @mroeschke 

merge on green.

**Comment by jorisvandenbossche** [comment]
New features are for 0.24.0, @mroeschke can you move the whatsnew?

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 27]
```diff
@@ -179,6 +179,55 @@ def test_to_datetime_format_weeks(self, cache):
...
+         ['2010-01-01 12:00:00 UTC'] * 2,
+         [pd.Timestamp('2010-01-01 12:00:00', tz='UTC')] * 2],
+        ['%Y-%m-%d %H:%M:%S %Z',
+         ['2010-01-01 12:00:00 UTC',
+          '2010-01-01 12:00:00 GMT',
+          '2010-01-01 12:00:00 US/Pacific'],
+         [pd.Timestamp('2010-01-01 12:00:00', tz='UTC'),
+          pd.Timestamp('2010-01-01 12:00:00', tz='GMT'),
+          pd.Timestamp('2010-01-01 12:00:00', tz='US/Pacific')]],
+        ['%Y-%m-%d %H:%M:%S %z',
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/datetimes/test_tools.py]: Can you one of them, eg this one, without the space before the tz?

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 28]
```diff
@@ -179,6 +179,55 @@ def test_to_datetime_format_weeks(self, cache):
...
+         ['2010-01-01 12:00:00 +0100', '2010-01-01 12:00:00 -0100'],
+         [pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(60)),
+          pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(-60))]],
+        ['%Y-%m-%d %H:%M:%S %z',
+         ['2010-01-01 12:00:00 Z', '2010-01-01 12:00:00 Z'],
+         [pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(0)),
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/datetimes/test_tools.py]: Should this be UTC or a fixed offset of 0 ?

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 29]
```diff
@@ -179,6 +179,55 @@ def test_to_datetime_format_weeks(self, cache):
...
+         [pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(60))] * 2],
+        ['%Y-%m-%d %H:%M:%S %z',
+         ['2010-01-01 12:00:00 +0100', '2010-01-01 12:00:00 -0100'],
+         [pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(60)),
+          pd.Timestamp('2010-01-01 12:00:00',
+                       tzinfo=pytz.FixedOffset(-60))]],
+        ['%Y-%m-%d %H:%M:%S %z',
+         ['2010-01-01 12:00:00 Z', '2010-01-01 12:00:00 Z'],
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/datetimes/test_tools.py]: Should this also work with `%Z`? 
It seems that with `datetime.datetime.strptime` it does not work with either

File: pandas/tests/indexes/datetimes/test_tools.py [hunk 30]
```diff
@@ -179,6 +179,55 @@ def test_to_datetime_format_weeks(self, cache):
...
+        # GH 13486
+        result = pd.to_datetime(dates, format=fmt, box=box)
+        expected = const(expected_dates)
+        getattr(tm, assert_equal)(result, expected)
+
+        with pytest.raises(ValueError):
+            pd.to_datetime(dates, format=fmt, box=box, utc=True)
+
+    @pytest.mark.parametrize('offset', [
+        '+0', '-1foo', 'UTCbar', ':10', '+01:000:01'])
```
Comment (jorisvandenbossche) [inline pandas/tests/indexes/datetimes/test_tools.py]: Can you add an empty string here as well?

**Review by jorisvandenbossche — COMMENTED** [review]
Nice addition! 
Added some comments

File: pandas/tests/indexes/datetimes/test_tools.py (hunk 28 above)
Comment (mroeschke) [reply pandas/tests/indexes/datetimes/test_tools.py]: pytz coerces a fixed offset of 0 to UTC

```
In [2]: pytz.FixedOffset(0)
Out[2]: <UTC>
```

But making it explicit here that %z should return `pytz.FixedOffset(0)`

File: pandas/tests/indexes/datetimes/test_tools.py (hunk 29 above)
Comment (mroeschke) [reply pandas/tests/indexes/datetimes/test_tools.py]: The regex I pulled from https://github.com/python/cpython/blob/master/Lib/_strptime.py has an option for 'Z' with` %z`:

https://github.com/python/cpython/blob/483000e164ec68717d335767b223ae31b4b720cf/Lib/_strptime.py#L204

But `%Z` only makes timezones found in the system local time available, i.e. no 'Z' option.

https://github.com/python/cpython/blob/483000e164ec68717d335767b223ae31b4b720cf/Lib/_strptime.py#L210-L212

Comment (jorisvandenbossche) [reply pandas/tests/indexes/datetimes/test_tools.py]: OK (that's probably a newer addition to python), then it makes sense to follow upstream python to be consistent

**Comment by jreback** [comment]
thanks @mroeschke nice patch!

betcha didn't think it would be this long when you first put it up! hahah. tests and code look great!

**Comment by mroeschke** [comment]
ha no problem, thanks!
